proof/courier.go

```diff
@@ -295,53 +359,297 @@ func (h *HashMailCourier) DeliverProof(ctx context.Context, addr address.Taro,
 	//
 	// TODO(roasbeef): ok that both sides might be on the same side here?
-	receiverStreamID := deriveReceiverStreamID(addr)
 	log.Infof("Creating receiver mailbox w/ sid=%x", receiverStreamID)
 	if err := h.mailbox.Init(ctx, receiverStreamID); err != nil {
-		return err
+		return fmt.Errorf("failed to init receiver ACK mailbox: %w",
+			err)
 	}
 
-	// We'll wait to receive the ACK from the remote party over their
-	// stream.
-	log.Infof("Waiting for receiver ACK via sid=%x", receiverStreamID)
-	if err := h.mailbox.RecvAck(ctx, receiverStreamID); err != nil {
-		return err
+	return nil
+}
+
+// timeSinceLastDeliveryAttempt calculates time duration which has elapsed since
+// the last delivery attempt.
+func timeSinceLastDeliveryAttempt(timestamps []time.Time) time.Duration {
+	// If there are no previous proof delivery attempts, then we'll
+	// return early.
+	if len(timestamps) == 0 {
+		return time.Duration(0)
 	}
 
-	log.Infof("Received ACK from receiver! Cleaning up mailboxes...")
+	// Otherwise we'll select the latest timestamp and compute the surpassed
+	// time relative to the current time.
 
-	// Once we receive this ACK, we can clean up our mailbox and also the
-	// receiver's mailbox.
-	if err := h.mailbox.CleanUp(ctx, senderStreamID); err != nil {
-		return err
+	// Get the latest timestamp without assuming order.
+	latestTimestamp := timestamps[0]
+	for _, timestamp := range timestamps {
+		if timestamp.After(latestTimestamp) {
+			latestTimestamp = timestamp
+		}
+	}
+
+	return time.Since(latestTimestamp)
+}
+
+// BackoffExecError is an error returned when the backoff execution fails.
+// This error wraps the underlying error returned by the execution function.
+// It allows the porter to determine whether the state machine should be halted
+// or not.
+type BackoffExecError struct {
+	execErr error
+}
+
+func (e *BackoffExecError) Error() string {
+	return fmt.Sprintf("backoff exec error: %s", e.execErr.Error())
+}
+
+// backoffExec attempts to execute the given `exec` function using a repeating
+// backoff time delayed strategy. The backoff strategy is used to ensure
+// that we don't spam the hashmail service with proof delivery attempts.
+func (h *HashMailCourier) backoffExec(ctx context.Context,
+	exec func() error) error {
+
+	var (
+		backoff          = h.cfg.BackoffCfg.InitialBackoff
+		numTries         = h.cfg.BackoffCfg.NumTries
+		maxBackoff       = h.cfg.BackoffCfg.MaxBackoff
+		err        error = nil
+	)
+
+	for i := 0; i < numTries; i++ {
+		// Don't delay at first try. Wait before reattempting
+		// delivery if a backoff duration is specified.
+		if i > 0 && backoff != 0 {
+			// Notify subscribers that backoff wait is about to
+			// commence.
+			transferEvent := NewReceiverProofBackoffWaitEvent(
+				backoff, int64(i),
+			)
+			h.publishSubscriberEvent(transferEvent)
+
+			err = h.wait(ctx, backoff)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = exec()
+		if err != nil {
```

Comment:
if you think it simplifie the logic a bit we can inverse this `if`

`if err == nil { 
    // we succeeded
    return (or if you prefer break)
}
// no need for else neither continue we can put the logic directly here
`

Also, I think the InitialBackoff is multiplied by 2 before we wait on it

Reply:
Added a fixup commit which reduces logging and simplifies the backoff loop. Let me know what you think!